planner/core/push_down_predicate.go

```diff
@@ -0,0 +1,246 @@
+	sort.SliceStable(exprGroups, func(i, j int) bool {
+		return exprGroups[i].selectivity < exprGroups[j].selectivity || (exprGroups[i].selectivity == exprGroups[j].selectivity && len(exprGroups[i].exprs) < len(exprGroups[j].exprs))
+	})
+
+	return exprGroups
+}
+
+/**
+ * @brief: This function is used to check if the conditions contain heavy cost functions.
+ * @param: conds: some conditions of PhysicalSelection to be checked
+ * @note: heavy cost functions are functions that may cause a lot of memory allocation or disk IO.
+ * 		  For example, JSON functions.(To be supplemented)
+ */
+func withHeavyCostFunction(conds []expression.Expression) bool {
+	for _, cond := range conds {
+		if binop, ok := cond.(*expression.ScalarFunction); ok {
+			switch binop.FuncName.L {
+			// JSON functions
+			case ast.JSONType, ast.JSONExtract, ast.JSONUnquote, ast.JSONValid, ast.JSONSearch, ast.JSONContains, ast.JSONContainsPath, ast.JSONKeys, ast.JSONLength, ast.JSONDepth, ast.JSONQuote, ast.JSONSet, ast.JSONInsert, ast.JSONReplace, ast.JSONRemove, ast.JSONMerge, ast.JSONArrayAppend, ast.JSONArrayInsert, ast.JSONMergePatch, ast.JSONMergePreserve, ast.JSONPretty, ast.JSONStorageSize:
+				return true
+				// TODO: add more heavy cost functions
+			}
+		}
+	}
+	return false
+}
+
+/**
+ * @brief: This function is used to remove the conditions that needed to be pushed down.
+ * @param: physicalSelection: the PhysicalSelection to be modified
+ * @param: exprs: the conditions to be removed
+ */
+func removeSpecificExprsFromSelection(physicalSelection *PhysicalSelection, exprs []expression.Expression) {
+	var newConds []expression.Expression
+	for _, cond := range physicalSelection.Conditions {
+		if !expression.Contains(exprs, cond) {
+			newConds = append(newConds, cond)
+		}
+	}
+	physicalSelection.Conditions = newConds
+}
+
+/**
+ * @brief: This function is used to push down the some filter conditions of the selection to the tablescan.
+ * @param: sctx: the session context
+ * @param: physicalSelection: the PhysicalSelection containing the conditions to be pushed down
+ * @param: physicalTableScan: the PhysicalTableScan to be pushed down to
+ */
+func predicatePushDownToTableScan(sctx sessionctx.Context, physicalSelection *PhysicalSelection, physicalTableScan *PhysicalTableScan) {
+	conds := physicalSelection.Conditions
+	if len(conds) == 0 {
+		return
+	}
+
+	// group the conditions by columns and sort them by selectivity
+	sortedConds := groupByColumnsSortBySelectivity(sctx, conds, physicalTableScan)
+
+	var selectedConds []expression.Expression
+	selectedIncome := 0.0
+	selectedColumnCount := 0
+	totalColumnCount := len(physicalTableScan.Columns)
+
+	for _, exprGroup := range sortedConds {
+		// If excuting light cost condition first,
+		// the rows needed to be exucuted by the heavy cost condition will be reduced.
+		// So if the group contains heavy cost functions, skip it.
+		if withHeavyCostFunction(exprGroup.exprs) {
```

Comment:
Can we check it in `groupByColumnsSortBySelectivity` before line 149 to reduce some sort compute?

Reply:
Good point!